adr/2025_03_10-suggestions/README.md

```diff
@@ -0,0 +1,178 @@
+  - Which user proposed the suggestion
+  - An optional thread identifier, for discussion around why the change was made
+
+**Prior Art:**
+
+- Google Docs
+- [Manuscripts Track Changes](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin), [prosemirror-suggest-changes](https://github.com/handlewithcarecollective/prosemirror-suggest-changes), [prosemirror-suggestion-mode](https://github.com/davefowler/prosemirror-suggestion-mode/) and [wax prosemirror tracked changes](https://gitlab.coko.foundation/wax/wax-prosemirror/-/blob/master/wax-prosemirror-core/src/utilities/track-changes/trackedTransaction.js?ref_type=heads)
+
+### Similarities
+
+Versioning and suggestions have quite a bit of overlap between them, they both rely on having a set of changes which store:
+
+- user: who made the change (the change's attribution)
+- timestamp: of when the change was made
+- content: what changed & it's location in the document
+
+Ideally, we could re-use some of the logic and data structures between them for both features, since they overlap in representation and visualization to users.
+
+## What we need to decide
+
+### How should Y.js snapshots be compared
+
+Y.js already has versioning in the form of snapshots of the Y.js document, at a point in time. In Y.js, the snapshots have the nice property of having some form of a history associated to those edits. To figure out what changed, you could calculate a diff between the two histories of the document. Doing a comparison based on the document content alone, would not meet our requirement of being able to display who made the change and when. Luckily, in Y.js, there is additional metadata that we can extract from Y.js which can tell us which `clientID` made the change (`clientID` being an ID a user assumes when making edits to a document). More on how to calculate that diff here: [Diff two Y.js Snapshots](./diff-yjs-snapshots.md)
+
+This would be the most straightforward way to capture changes across versions of a document. The question then becomes how to display those changes within a prosemirror-based editor, which sort of depends on the approach taken for suggestions.
+
+### How should suggestions be stored
+
+Suggestions are different from versioning in that, they are not a history of the document, but a set of pending changes that can be applied to the document. They may be rejected, or may be applied to the document. So, in a sense, they are a patch to the document. But, there are a few different mental models we can use to represent or store these changes to the document:
+
+- [Seeing a suggestion as a _branch of the document history_, and the suggestion as being the diff between the two branches](./suggestions-as-history.md)
+- [Storing document _patches_ and later applying that patch to a document](./suggestions-as-patches.md)
+- [Suggestions as _document content_ which stores the changes within the document](./suggestions-as-content.md)
+
+### AttributionManager
```

Comment:
Kevin will write more about the attribution manager, but to give a summary of how it will work:

 - The AttributionManager will sit on the server and be able to attribute updates & elements in a trusted store (the server), to specific users. Those users will be atrtributed to specific clientIDs & clocks.
 - Technically, a user can re-use another's, clientID, so the attribution manager must take care to actually known at what clock the change was made, to correctly identify the user.
 - The attribution manager will likely be implemented in a network-agnostic way similar to the awareness protocol, so that clients can receive the attribution of users from the server.